src/main.cpp

```diff
@@ -2528,7 +2568,58 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
     blockundo.vtxundo.reserve(block.vtx.size() - 1);
     int nChecked = 0;
     int nOrphansChecked = 0;
-    LOCK(cs_xval);
+    arith_uint256 nStartingChainWork = chainActive.Tip()->nChainWork;
+
+    // Create a vector for storing hashes that will be deleted from the unverified and perverified txn sets.
+    // We will delete these hashes only if and when this block is the one that is accepted saving us the unnecessary 
+    // repeated locking and unlocking of cs_xval.
+    std::vector<uint256> vHashesToDelete;
+
+    // Create a temporary view of the UTXO set
+    CCoinsViewCache viewTempCache(pcoinsTip);
+
+    // Section for boost scoped lock on the scriptcheck_mutex
+    boost::thread::id this_id(boost::this_thread::get_id());
+    {
+
+    // Get the next available mutex and then use it to find the associated scriptcheckqueue. Then lock this thread
+    // with the mutex so other thread can used this scriptcheckqueue until all inputs have been checked and the 
+    // scriptcheckqueue threads have all returned.
+    boost::shared_ptr<boost::mutex> scriptcheck_mutex = allScriptCheckQueues.GetScriptCheckMutex();
+    if (fParallel) cs_main.unlock(); // unlock cs_main, we may be waiting here for a while before aquiring the scoped lock below
+    boost::mutex::scoped_lock lock(*scriptcheck_mutex);
+
+    // Get the next available scriptcheckqueue.
+    CCheckQueue<CScriptCheck>* pScriptQueue = allScriptCheckQueues.GetScriptCheckQueue(scriptcheck_mutex);
+    assert(pScriptQueue != NULL);
+
+    //CCheckQueueControl<CScriptCheck> control(fScriptChecks && nScriptCheckThreads ? &scriptcheckqueue : NULL);
+    CCheckQueueControl<CScriptCheck> control(fScriptChecks && nScriptCheckThreads ? pScriptQueue : NULL);
+    {
+    LOCK(cs_blockvalidationthread);
+    if (fParallel) {
+        // We need to place a Quit here because we do not want to assign a script queue to a thread of activity
+        // if another thread has just won the race and has sent an fQuit.
+        if (mapBlockValidationThreads[this_id].fQuit) {
+            LogPrint("parallel", "fQuit 0 called - Stopping validation of this block and returning\n");
+            cs_main.lock(); // must lock before returning.
```

Comment:
cs_main must be locked when we return as it is a recursive mutex and must be unlocked by the caller.